drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1031 @@
+	} else {
+		secur_reg &= ~SPINAND_SECURE_BIT_CONT;
+	}
+	ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, secur_reg);
+	if (ret != 0) {
+		LOG_ERR("set feature failed: %d", ret);
+		goto out;
+	}
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out;
+	}
+	if ((secur_reg & SPINAND_SECURE_BIT_CONT) == 0) {
+		LOG_ERR("Enable continuous read failed: %d\n", secur_reg);
+	}
+
+out:
+	release_device(dev);
+	return ret;
+}
+
+static int spi_nand_read_cont(const struct device *dev, off_t addr, void *dest, size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+
+	acquire_device(dev);
+	ret = spi_nand_access(dev, SPI_NAND_CMD_PAGE_READ,
+			      NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+			      addr >> data->page_shift, NULL, 0);
+	if (ret != 0) {
+		LOG_ERR("page read failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_wait_until_ready(dev);
+	if (ret != 0) {
+		LOG_ERR("wait ready failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_access(dev, SPI_NAND_CMD_READ_CACHE,
+			      NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+			      ((addr >> data->page_shift) & SPI_NAND_PAGE_MASK), dest, size);
+	if (ret != 0) {
+		LOG_ERR("read from cache failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_conti_read_enable(dev, false);
+
+out:
+	release_device(dev);
+
+	return ret;
+}
+
+static int spi_nand_read_normal(const struct device *dev, off_t addr, void *dest, size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+	uint32_t offset = 0;
+	uint32_t chunk = 0;
+
+	acquire_device(dev);
+	while (size > 0) {
+		/* Read on _page_size_bytes boundaries (Default 2048 bytes a page) */
+		offset = addr % data->page_size;
+		chunk = (offset + size < data->page_size) ? size : (data->page_size - offset);
+		ret = spi_nand_access(dev, SPI_NAND_CMD_PAGE_READ,
+				      NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+				      addr >> data->page_shift, NULL, 0);
+		if (ret != 0) {
+			LOG_ERR("page read failed: %d", ret);
+			goto out;
+		}
+
+		ret = spi_nand_wait_until_ready(dev);
+		if (ret != 0) {
+			LOG_ERR("wait ready failed: %d", ret);
+			goto out;
+		}
+
+		ret = spi_nand_access(dev, SPI_NAND_CMD_READ_CACHE,
+				      NAND_ACCESS_ADDRESSED | NAND_ACCESS_16BIT_ADDR |
+					      NAND_ACCESS_DUMMY,
+				      (addr & SPI_NAND_PAGE_MASK), dest, chunk);
+		if (ret != 0) {
+			LOG_ERR("read from cache failed: %d", ret);
+			goto out;
+		}
+
+		dest = (uint8_t *)dest + chunk;
+
+		addr = (addr + SPI_NAND_PAGE_OFFSET) & (~SPI_NAND_PAGE_MASK);
+		size -= chunk;
+	}
+
+out:
+	release_device(dev);
+
+	return ret;
+}
+
+static int spi_nand_read(const struct device *dev, off_t addr, void *dest, size_t size)
+{
+	int ret = 0;
+	struct spi_nand_data *data = dev->data;
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > data->flash_size)) {
+		return -EINVAL;
+	}
```

Comment:
With `size` == 0 we are supposed to just return 0.
Additionally, to be safe against int overflows, the condition should be:
```
if ((addr < 0) || (addr >= data->flash_size) || (size > data->flash_size) || ((data->flash_size - addr) < size))
```